In [1]:
import  pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures # 다항 속성 
from sklearn.metrics import mean_squared_error

In [3]:
df = pd.read_csv("boston.csv") # 목표 변수 = MEDV
df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

In [10]:
x = df.iloc[:,:-1]
x

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [16]:
x.shape

(506, 13)

In [11]:
y = df[["MEDV"]]
y

MEDV
0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
..    ...
501  22.4
502  20.6
503  23.9
504  22.0
505  11.9

[506 rows x 1 columns]

In [12]:
polynomial_transformer = PolynomialFeatures(2) # 다항 변형기 생성 

In [13]:
polynomial_data = polynomial_transformer.fit_transform(x)
polynomial_data

array([[1.00000000e+00, 6.32000000e-03, 1.80000000e+01, ...,
        1.57529610e+05, 1.97656200e+03, 2.48004000e+01],
       [1.00000000e+00, 2.73100000e-02, 0.00000000e+00, ...,
        1.57529610e+05, 3.62766600e+03, 8.35396000e+01],
       [1.00000000e+00, 2.72900000e-02, 0.00000000e+00, ...,
        1.54315409e+05, 1.58310490e+03, 1.62409000e+01],
       ...,
       [1.00000000e+00, 6.07600000e-02, 0.00000000e+00, ...,
        1.57529610e+05, 2.23851600e+03, 3.18096000e+01],
       [1.00000000e+00, 1.09590000e-01, 0.00000000e+00, ...,
        1.54802902e+05, 2.54955600e+03, 4.19904000e+01],
       [1.00000000e+00, 4.74100000e-02, 0.00000000e+00, ...,
        1.57529610e+05, 3.12757200e+03, 6.20944000e+01]])

In [14]:
polynomial_data.shape

(506, 105)

위에서는 506, 13이 어떻게?  
2 차원이라는 가정하에
- x1**2, x2**2 -> 13개
- x1*x2, x1*x3 ... x12*x13 -> 78 => 2차항 91개
- 1 차항 -> 13개
- 0 차항 -> 1개
- total 105개 

In [20]:
polynomial_transformer.get_feature_names_out(x.columns) # 가능한 모든 2차 조합을 알려줌

array(['1', 'CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS',
       'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'CRIM^2', 'CRIM ZN',
       'CRIM INDUS', 'CRIM CHAS', 'CRIM NOX', 'CRIM RM', 'CRIM AGE',
       'CRIM DIS', 'CRIM RAD', 'CRIM TAX', 'CRIM PTRATIO', 'CRIM B',
       'CRIM LSTAT', 'ZN^2', 'ZN INDUS', 'ZN CHAS', 'ZN NOX', 'ZN RM',
       'ZN AGE', 'ZN DIS', 'ZN RAD', 'ZN TAX', 'ZN PTRATIO', 'ZN B',
       'ZN LSTAT', 'INDUS^2', 'INDUS CHAS', 'INDUS NOX', 'INDUS RM',
       'INDUS AGE', 'INDUS DIS', 'INDUS RAD', 'INDUS TAX',
       'INDUS PTRATIO', 'INDUS B', 'INDUS LSTAT', 'CHAS^2', 'CHAS NOX',
       'CHAS RM', 'CHAS AGE', 'CHAS DIS', 'CHAS RAD', 'CHAS TAX',
       'CHAS PTRATIO', 'CHAS B', 'CHAS LSTAT', 'NOX^2', 'NOX RM',
       'NOX AGE', 'NOX DIS', 'NOX RAD', 'NOX TAX', 'NOX PTRATIO', 'NOX B',
       'NOX LSTAT', 'RM^2', 'RM AGE', 'RM DIS', 'RM RAD', 'RM TAX',
       'RM PTRATIO', 'RM B', 'RM LSTAT', 'AGE^2', 'AGE DIS', 'AGE RAD',
       'AGE TAX', 'AGE PTRATIO', 'AGE 

In [21]:
len(polynomial_transformer.get_feature_names_out(x.columns)) # 105개인 거 확인 

105

새로운 X를 만든 것

In [23]:
x = pd.DataFrame(polynomial_data, columns=polynomial_transformer.get_feature_names_out(x.columns))
x

1     CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  ...  \
0    1.0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  ...   
1    1.0  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  ...   
2    1.0  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  ...   
3    1.0  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  ...   
4    1.0  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  ...   
..   ...      ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  1.0  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  ...   
502  1.0  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  ...   
503  1.0  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  ...   
504  1.0  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  ...   
505  1.0  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  ...   

       TAX^2  TAX PTRATIO      TAX B  TAX LSTAT  PTRATIO^2  PTRATIO B  \
0    87616.0       4528.8  117482.40    1474.08     234.09   6072.570   
1    58564.0       4307.6   96049.80    2211.88     316.84   7064.820   
2    58564.0       4307.6   95064.86     975.26     316.84   6992.374   
3    49284.0       4151.4   87607.86     652.68     349.69   7379.581   
4    49284.0       4151.4   88111.80    1183.26     349.69   7422.030   
..       ...          ...        ...        ...        ...        ...   
501  74529.0       5733.0  107013.27    2639.91     441.00   8231.790   
502  74529.0       5733.0  108353.70    2478.84     441.00   8334.900   
503  74529.0       5733.0  108353.70    1539.72     441.00   8334.900   
504  74529.0       5733.0  107411.85    1769.04     441.00   8262.450   
505  74529.0       5733.0  108353.70    2151.24     441.00   8334.900   

     PTRATIO LSTAT          B^2    B LSTAT  LSTAT^2  
0           76.194  157529.6100  1976.5620  24.8004  
1          162.692  157529.6100  3627.6660  83.5396  
2           71.734  154315.4089  1583.1049  16.2409  
3           54.978  155732.8369  1160.2122   8.6436  
4           99.671  157529.6100  2115.4770  28.4089  
..             ...          ...        ...      ...  
501        203.070  153656.1601  3790.5433  93.5089  
502        190.680  157529.6100  3603.8520  82.4464  
503        118.440  157529.6100  2238.5160  31.8096  
504        136.080  154802.9025  2549.5560  41.9904  
505        165.480  157529.6100  3127.5720  62.0944  

[506 rows x 105 columns]

In [24]:
# 모델 나누기  
X_train, X_test, y_train, y_test = train_test_split(x,y, test_size = 0.2, random_state=5)

In [25]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(404, 105)
(102, 105)
(404, 1)
(102, 1)


In [27]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [28]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [29]:
model.coef_ # theta1

array([[-2.55720149e-07, -5.09146959e+00, -1.65753983e-01,
        -5.97358605e+00,  2.43179259e+01,  1.65180559e+02,
         2.19910116e+01,  1.03167123e+00, -5.66895775e+00,
         3.22443249e+00, -1.10055942e-02,  5.35127787e+00,
        -4.81524409e-02,  7.53109325e-01,  2.16774682e-03,
         2.69938772e-01,  5.87901385e-01,  2.41731932e+00,
        -2.52413195e-02,  8.92859572e-02, -5.18832420e-03,
        -5.77807152e-02,  3.55602049e-01, -3.86092282e-02,
         5.43572101e-01, -3.18134358e-04,  2.40035425e-02,
        -7.48850220e-04, -7.16133310e-03, -1.06886010e-01,
        -1.27782609e+00,  2.50137719e-02,  1.14111417e-04,
        -1.25254119e-02, -4.68024813e-03,  6.05725185e-04,
        -8.57873132e-03,  1.85030053e-03, -4.64730601e-03,
         3.08484808e-02, -2.09065897e-01,  1.30035723e+00,
         3.13497405e-01,  6.72540163e-04,  7.51823883e-02,
        -7.38014890e-03,  4.23364348e-04, -6.72155118e-03,
         6.42107774e-03, -5.32275093e-03,  2.43179262e+0

In [30]:
model.intercept_ # theta0

array([-141.89855578])

In [31]:
# 평가는 test set
y_test_prediction = model.predict(X_test)
y_test_prediction

array([[42.00919669],
       [28.22372855],
       [25.52688254],
       [ 9.05454749],
       [33.5903981 ],
       [10.49656654],
       [23.0597188 ],
       [30.35330832],
       [24.2246103 ],
       [22.30153137],
       [33.11791368],
       [20.74987646],
       [20.19664017],
       [32.37012974],
       [27.35813074],
       [20.46264673],
       [13.68533393],
       [12.51890372],
       [15.88730519],
       [12.47719001],
       [ 3.72827178],
       [20.49819423],
       [44.0935012 ],
       [23.31624023],
       [33.2791559 ],
       [ 9.43400657],
       [24.71325022],
       [21.79459244],
       [24.06481669],
       [27.42603119],
       [15.3289399 ],
       [ 6.80742071],
       [16.76243454],
       [13.13446141],
       [25.10746984],
       [22.92666537],
       [29.58310464],
       [10.66362649],
       [47.75889196],
       [35.24353036],
       [19.90943076],
       [15.25028015],
       [28.15702648],
       [14.02415348],
       [26.11171521],
       [28

In [32]:
# 8.236881612652455 선형 회귀 결과

In [34]:
mean_squared_error(y_test, y_test_prediction, squared=False) 
# 선형 회귀에 비해 많이 줄었음 

3.196527651588392